File: src/project_exporter_desktop/reports/insights/file_statistics.py
from __future__ import annotations

from collections import Counter
from pathlib import Path
from typing import Any

from ...utils.inventory import is_non_ascii, path_depth
from ...utils.path_utils import rel_display
from ...utils.text_utils import format_bytes
from ...utils.time_utils import human_now


def write_file_statistics_report(
    copied_root: Path, output_file: Path, inventory: dict[str, Any]
) -> None:
    files: list[Path] = inventory["files"]
    sizes: list[tuple[Path, int]] = inventory["sizes"]
    by_ext_count: Counter[str] = inventory["by_ext_count"]
    by_ext_size: Counter[str] = inventory["by_ext_size"]

    empty_files = [(p, s) for p, s in sizes if s == 0]
    spaced = [p for p in files if " " in p.name]
    non_ascii = [p for p in files if is_non_ascii(str(p.relative_to(copied_root)))]
    deepest = sorted(files, key=lambda p: path_depth(p, copied_root), reverse=True)[:25]
    largest = sorted(sizes, key=lambda item: item[1], reverse=True)[:25]
    long_paths = [p for p in files if len(str(p)) >= 240]

    with output_file.open("w", encoding="utf-8", newline="\n", errors="replace") as out:
        out.write("=== File Statistics ===\n")
        out.write(f"Generated: {human_now()}\n")
        out.write("=" * 100 + "\n\n")

        out.write("--- Files by extension ---\n")
        out.write(f"{'Extension':<20} {'Files':>10} {'Total size':>14}\n")
        out.write(f"{'-' * 20} {'-' * 10} {'-' * 14}\n")
        for ext, count in by_ext_count.most_common():
            out.write(f"{ext:<20} {count:>10,} {format_bytes(by_ext_size[ext]):>14}\n")

        out.write("\n--- Top 25 largest files ---\n")
        for path, size in largest:
            out.write(f"{format_bytes(size):>12}  {rel_display(path, copied_root)}\n")

        out.write("\n--- Top 25 deepest files ---\n")
        for path in deepest:
            out.write(
                f"depth={path_depth(path, copied_root):>2}  {rel_display(path, copied_root)}\n"
            )

        out.write("\n--- Empty files ---\n")
        if empty_files:
            for path, _size in empty_files[:100]:
                out.write(f"{rel_display(path, copied_root)}\n")
            if len(empty_files) > 100:
                out.write(f"... and {len(empty_files) - 100:,} more\n")
        else:
            out.write("None detected.\n")

        out.write("\n--- Files with spaces in names ---\n")
        if spaced:
            for path in spaced[:100]:
                out.write(f"{rel_display(path, copied_root)}\n")
            if len(spaced) > 100:
                out.write(f"... and {len(spaced) - 100:,} more\n")
        else:
            out.write("None detected.\n")

        out.write("\n--- Files with non-ASCII paths ---\n")
        if non_ascii:
            for path in non_ascii[:100]:
                out.write(f"{rel_display(path, copied_root)}\n")
            if len(non_ascii) > 100:
                out.write(f"... and {len(non_ascii) - 100:,} more\n")
        else:
            out.write("None detected.\n")

        out.write("\n--- Potential Windows long-path risks >= 240 characters ---\n")
        if long_paths:
            for path in long_paths[:100]:
                out.write(f"{len(str(path)):>4} chars  {rel_display(path, copied_root)}\n")
            if len(long_paths) > 100:
                out.write(f"... and {len(long_paths) - 100:,} more\n")
        else:
            out.write("None detected.\n")

File: crates/codepack-reports/src/reports/file_statistics.rs
//! `02_file_statistics.txt`, ported from legacy
//! `reports/insights/file_statistics.py::write_file_statistics_report`.

use std::path::Path;

use codepack_tokens::format_bytes;

use crate::context::ReportContext;
use crate::error::ReportError;
use crate::paths::path_depth;
use crate::plugin::ReportJob;
use crate::profile;

pub const JOB: ReportJob = ReportJob {
    filename: "02_file_statistics.txt",
    profiles: profile::FILE_STATISTICS_TXT,
    description: "File counts by extension, deepest paths, empty / suspicious files.",
    run: write_file_statistics_report,
};

const LIST_LIMIT: usize = 100;
/// Legacy: "conservative; Windows `MAX_PATH` is 260, but tools like `robocopy` fail earlier."
const LONG_PATH_THRESHOLD: usize = 240;

fn is_non_ascii(text: &str) -> bool {
    text.chars().any(|ch| ch as u32 > 127)
}

fn write_list(out: &mut String, title: &str, items: &[String]) {
    out.push_str(&format!("\n--- {title} ---\n"));
    if items.is_empty() {
        out.push_str("None detected.\n");
        return;
    }
    for item in items.iter().take(LIST_LIMIT) {
        out.push_str(item);
        out.push('\n');
    }
    if items.len() > LIST_LIMIT {
        out.push_str(&format!("... and {} more\n", items.len() - LIST_LIMIT));
    }
}

fn write_file_statistics_report(
    ctx: &ReportContext<'_>,
    output_file: &Path,
) -> Result<(), ReportError> {
    let inventory = ctx.inventory;

    let mut sorted_by_size: Vec<&crate::context::InventoryFile> = inventory.files.iter().collect();
    sorted_by_size.sort_by_key(|file| std::cmp::Reverse(file.size));

    let mut sorted_by_depth: Vec<&crate::context::InventoryFile> = inventory.files.iter().collect();
    sorted_by_depth.sort_by_key(|file| std::cmp::Reverse(path_depth(&file.relative_path)));

    let empty_files: Vec<String> = inventory
        .files
        .iter()
        .filter(|file| file.size == 0)
        .map(|file| file.relative_path.clone())
        .collect();
    let spaced: Vec<String> = inventory
        .files
        .iter()
        .filter(|file| file.relative_path.contains(' '))
        .map(|file| file.relative_path.clone())
        .collect();
    let non_ascii: Vec<String> = inventory
        .files
        .iter()
        .filter(|file| is_non_ascii(&file.relative_path))
        .map(|file| file.relative_path.clone())
        .collect();
    let long_paths: Vec<(String, usize)> = inventory
        .files
        .iter()
        .filter(|file| file.relative_path.len() >= LONG_PATH_THRESHOLD)
        .map(|file| (file.relative_path.clone(), file.relative_path.len()))
        .collect();

    let mut out = String::new();
    out.push_str("=== File Statistics ===\n");
    out.push_str(&format!("Generated: {}\n", ctx.plan.generated_at));
    out.push_str(&"=".repeat(100));
    out.push_str("\n\n");

    out.push_str("--- Files by extension ---\n");
    out.push_str(&format!(
        "{:<20} {:>10} {:>14}\n",
        "Extension", "Files", "Total size"
    ));
    out.push_str(&format!(
        "{} {} {}\n",
        "-".repeat(20),
        "-".repeat(10),
        "-".repeat(14)
    ));
    for stat in &inventory.by_extension {
        out.push_str(&format!(
            "{:<20} {:>10} {:>14}\n",
            stat.extension,
            stat.count,
            format_bytes(stat.total_size)
        ));
    }

    out.push_str("\n--- Top 25 largest files ---\n");
    for file in sorted_by_size.iter().take(25) {
        out.push_str(&format!(
            "{:>12}  {}\n",
            format_bytes(file.size),
            file.relative_path
        ));
    }

    out.push_str("\n--- Top 25 deepest files ---\n");
    for file in sorted_by_depth.iter().take(25) {
        out.push_str(&format!(
            "depth={:>2}  {}\n",
            path_depth(&file.relative_path),
            file.relative_path
        ));
    }

    write_list(&mut out, "Empty files", &empty_files);
    write_list(&mut out, "Files with spaces in names", &spaced);
    write_list(&mut out, "Files with non-ASCII paths", &non_ascii);

    out.push_str("\n--- Potential Windows long-path risks >= 240 characters ---\n");
    if long_paths.is_empty() {
        out.push_str("None detected.\n");
    } else {
        for (path, length) in long_paths.iter().take(LIST_LIMIT) {
            out.push_str(&format!("{length:>4} chars  {path}\n"));
        }
        if long_paths.len() > LIST_LIMIT {
            out.push_str(&format!("... and {} more\n", long_paths.len() - LIST_LIMIT));
        }
    }

    std::fs::write(output_file, out).map_err(|source| ReportError::Write {
        path: output_file.to_path_buf(),
        source,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_support::Fixture;

    #[test]
    fn lists_empty_spaced_and_extension_breakdown() {
        let fixture = Fixture::new(|root| {
            std::fs::write(root.join("empty.txt"), "").unwrap();
            std::fs::write(root.join("has space.txt"), "x").unwrap();
            std::fs::create_dir_all(root.join("src")).unwrap();
            std::fs::write(root.join("src/main.py"), "x = 1\n").unwrap();
        });
        let ctx = fixture.context("full");
        let out_dir = tempfile::tempdir().unwrap();
        let output_file = out_dir.path().join(JOB.filename);

        write_file_statistics_report(&ctx, &output_file).unwrap();

        let content = std::fs::read_to_string(&output_file).unwrap();
        assert!(content.starts_with("=== File Statistics ==="));
        assert!(content.contains("empty.txt"));
        assert!(content.contains("has space.txt"));
        assert!(content.contains("py"));
    }

    #[test]
    fn reports_none_detected_when_no_suspicious_files_exist() {
        let fixture = Fixture::new(|root| {
            std::fs::write(root.join("main.py"), "x = 1\n").unwrap();
        });
        let ctx = fixture.context("full");
        let out_dir = tempfile::tempdir().unwrap();
        let output_file = out_dir.path().join(JOB.filename);

        write_file_statistics_report(&ctx, &output_file).unwrap();

        let content = std::fs::read_to_string(&output_file).unwrap();
        assert!(content.contains("--- Empty files ---\nNone detected."));
    }
}
